src/consensus.rs

```diff
@@ -822,92 +760,228 @@ fn handle_committed_entries(
+        while let Some(message) = self.messages.recv().await {
+            self.send(&message).await;
+        }
     }
-    let message = &GrpcRaftMessage { message: bytes };
-
-    let result = transport_channel_pool
-        .with_channel_timeout(
-            &address,
-            |channel| async move {
-                let mut client = RaftClient::new(channel);
-                let mut request = tonic::Request::new(message.clone());
-                request.set_timeout(timeout);
-                client.send(request).await
-            },
-            Some(timeout),
-            0,
-        )
-        .await;
-
-    if is_snapshot {
-        // Should we ignore the error? Seems like it will only produce noise.
-        //
-        // - `send_message` is only called by the sub-task spawned by the consnsus thread.
-        // - `report_snapshot` sends a message back to the consensus thread.
-        // - It can only fail, if the "receiver" end of the channel is closed.
-        // - Which means consensus thread either resolved successfully, or failed.
-        // - So, if the consensus thread is shutting down, no need to log a misleading error...
-        // - ...or, if the consensus thread failed, then we should already have an error,
-        //   and it will only produce more noise.
-
-        let res = store.report_snapshot(
-            peer_id,
-            if result.is_ok() {
-                SnapshotStatus::Finish
+
+    async fn send(&mut self, message: &RaftMessage) {
+        if let Err(err) = self.try_send(message).await {
+            let peer_id = message.to;
+
+            if log::max_level() >= log::Level::Debug {
+                log::error!("Failed to send Raft message {message:?} to peer {peer_id}: {err}");
             } else {
-                SnapshotStatus::Failure
-            },
+                log::error!("Failed to send Raft message to peer {peer_id}: {err}");
+            }
+        }
+    }
+
+    async fn try_send(&mut self, message: &RaftMessage) -> anyhow::Result<()> {
+        let peer_id = message.to;
+
+        let uri = self.uri(peer_id).await?;
+
+        let mut bytes = Vec::new();
+        RaftMessage::encode(message, &mut bytes).context("failed to encode Raft message")?;
+        let grpc_message = GrpcRaftMessage { message: bytes };
+
+        let timeout = Duration::from_millis(
+            self.consensus_config.message_timeout_ticks * self.consensus_config.tick_period_ms,
         );
 
-        if let Err(err) = res {
-            log::error!("{}", err);
+        let res = self
+            .transport_channel_pool
+            .with_channel_timeout(
+                &uri,
+                |channel| async {
+                    let mut client = RaftClient::new(channel);
+                    let mut request = tonic::Request::new(grpc_message.clone());
+                    request.set_timeout(timeout);
+                    client.send(request).await
+                },
+                Some(timeout),
```

Comment:
Shouldn't we use a different timeout for hearbeat messages? The final conclusion on this is not clear to me after our discussion.

Reply:
I'm pondering on it right now, actually. On one hand, it kinda makes sense to make this timeout lower for heartbeat messages... but we also don't want it to be too small, cause if network conditions are bad, we risk not being able to send _any_ heartbeat at all (and, IIRC, we did hit this issue before). I think we may want to leave it as is for now, and try to come up with a more solid strategy for heartbeats in following PRs.